factor/email/classes/factor.php

```diff
@@ -121,4 +112,119 @@ public function get_all_user_factors() {
         $record['id'] = $DB->insert_record('tool_mfa', $record, true);
         return [(object) $record];
     }
+
+    public function has_input() {
+        if (self::is_ready()) {
+            return true;
+        }
+        return false;
+    }
+
+    public function get_state() {
+        if (!self::is_ready()) {
+            return \tool_mfa\plugininfo\factor::STATE_NEUTRAL;
+        }
+
+        return parent::get_state();
+    }
+
+    private static function is_ready() {
+        global $DB, $USER;
+
+        if (empty($USER->email)) {
+            return false;
+        }
+        if (!validate_email($USER->email)) {
+            return false;
+        }
+        if (over_bounce_threshold($USER)) {
+            return false;
+        }
+
+        // If this factor is revoked, set to not ready.
+        if ($DB->record_exists('tool_mfa', array('userid' => $USER->id, 'factor' => 'email', 'revoked' => 1))) {
+            return false;
+        }
+        return true;
+    }
+
+    private function generate_and_email_code() {
+        global $DB, $USER;
+
+        // Get instance that isnt parent email type (label check).
```

Comment:
Can you please add a comment here explaining the two types of records the email check